openmc/lattice.py

```diff
@@ -753,6 +764,336 @@ def is_valid_index(self, idx):
+            are represented as follows [top left, top, top right, left, right,
+            bottom left, bottom, bottom right]
+        attribute : String
+            Which universe attribute to use when detecting neighbor symmetry
+            patterns. Can be the 'id' or the 'name' of the lattice universes
+
+        """
+
+        # Check routine inputs
+        if self.ndim == 3:
+            raise NotImplementedError("LNS discretization is not implemented "
+                                      "for 3D lattices")
+
+        cv.check_value('strategy', strategy, ('degenerate', 'lns'))
+        cv.check_type('rotate_universe_with_neighbors',
+                      rotate_universe_with_neighbors, bool)
+        cv.check_type('universes_to_ignore', universes_to_ignore, Iterable,
+                      openmc.Universe)
+        cv.check_type('materials_to_clone', materials_to_clone, Iterable,
+                      openmc.Material)
+        cv.check_value('number of lattice_neighbors', len(lattice_neighbors), (0, 8))
+        cv.check_value('attribute', attribute, ('id', 'name'))
+
+        if len(lattice_neighbors) > 0:
+            if attribute == 'id':
+                 cv.check_type('lattice_neighbors', lattice_neighbors,
+                               Iterable, Integral)
+            elif attribute == 'name':
+                 cv.check_type('lattice_neighbors', lattice_neighbors,
+                               Iterable, str)
+
+        # Use outer universe if neighbors are missing and outer is defined
+        if self.outer is not None and len(lattice_neighbors) == 0:
+            lattice_neighbors = [getattr(self.outer, attribute) for i in range(8)]
+
+        # Create dummy neighbor for when the neighbor is not known at the edge
+        if len(lattice_neighbors) == 0:
+            if attribute == 'id':
+                dummy_neighbor = -1
+            elif attribute == 'name':
+                dummy_neighbor = "dummy"
+
+        # Dictionary that will keep track of where each pattern appears, how
+        # it was rotated and/or symmetrized
+        patterns = {}
+
+        # Analyze lattice, find unique patterns, ignoring rotations and symmetries
+        for j in range(self.shape[1]):
+            for i in range(self.shape[0]):
+
+                # Skip universes to ignore
+                if self.universes[j][i] in universes_to_ignore:
+                    continue
+
+                # Form the pattern at and around each universe
+                if attribute == "id":
+                    pattern = np.empty(shape=(3, 3), dtype=np.int)
+                if attribute == "name":
+                    pattern = np.empty(shape=(3, 3), dtype=np.dtype('U100'))
+                    if len(self.universes[j][i].name) > 100:
+                        warning.warn("Universe name is larger than 100 "
+                             "characters, which may hinder neighbor search.")
```

Comment:
That would solve the issue but then I probably would need to hash the result of getattr as well. Even when using the id attribute. It'll make the code harder to debug if we ever need to

Reply:
Okay, then maybe just use `dtype='O'`. That will work for a string of any length, and you won't need to have these different `if` cases for `id` versus `name`.

Also, you can move this initialization of `pattern` outside of the `for` loops so that it's only done once.